### PR TITLE
feat(eslint-plugin): [max-params] add function overload and function type support

### LINKED ISSUE #10293: Enhancement: [max-params] add option to check overload signatures as well
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/max-params/

### Description

Currently the `@typescript-eslint/max-params` rule only checks implementation signatures. I would like for an option to also check overload signatures which may have a different number of parameters.

### Fail

```TypeScript
// ❌ Function 'myFunc' has too many parameters (7). Maximum allowed is 4.
myFunc (a, b, c, d, e, f, g);

myFunc (...args) {
  // ...
}
```

### Pass

```TypeScript
myFunc (a, b, c);

myFunc (...args) {
  // ...
}
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by developer-bandi**
## PR Checklist

- [x] Addresses an existing open issue: fixes #10293
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Added support for the TSDeclareFunction and TSFunctionType types.

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/max-params.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/max-params.test.ts:1]: [Testing] These are reasonable `invalid` cases, but they should have corresponding `valid` ones too. Could you add those in please?

File: packages/eslint-plugin/tests/rules/max-params.test.ts [hunk 1]
```diff
@@ -98,5 +98,19 @@ class Foo {
       `,
       errors: [{ messageId: 'exceed' }],
     },
+    {
+      code: `
+function makeDate(m: number, d: number, y: number): Date;
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/max-params.test.ts:103]:
```suggestion
declare function makeDate(m: number, d: number, y: number): Date;
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Good start 🚀

**Review by JoshuaKGoldberg — APPROVED** [review]
🙌 looks great to me, thanks as always @developer-bandi!

**Review by kirkwaiblinger — APPROVED** [review]

**Pull request merged**